contracts/psp22/lib.rs

```diff
@@ -0,0 +1,153 @@
+#![cfg_attr(not(feature = "std"), no_std, no_main)]
+
+#[ink::contract]
+mod psp22 {
+
+    use ink::{codegen::EmitEvent, reflect::ContractEventBase, storage::Mapping};
+    use psp22_traits::{PSP22Error, PSP22};
+
+    #[ink(event)]
+    pub struct Approval {
+        #[ink(topic)]
+        owner: AccountId,
+        #[ink(topic)]
+        spender: AccountId,
+        value: Balance,
+    }
+
+    #[ink(event)]
+    pub struct Transfer {
+        #[ink(topic)]
+        from: Option<AccountId>,
+        #[ink(topic)]
+        to: Option<AccountId>,
+        value: Balance,
+    }
+
+    #[ink(storage)]
+    #[derive(Default)]
+    pub struct Token {
+        total_supply: Balance,
+        balances: Mapping<AccountId, Balance>,
+        allowances: Mapping<(AccountId, AccountId), Balance>,
+    }
+
+    pub type Event = <Token as ContractEventBase>::Type;
+
+    impl Token {
+        #[ink(constructor)]
+        pub fn new(total_supply: Balance) -> Self {
+            let mut balances = Mapping::default();
+            let caller = Self::env().caller();
+            balances.insert(caller, &total_supply);
+            Self {
+                total_supply,
+                balances,
+                allowances: Default::default(),
+            }
+        }
+
+        fn transfer_from_to(
+            &mut self,
+            from: &AccountId,
+            to: &AccountId,
+            value: Balance,
+        ) -> Result<(), PSP22Error> {
+            let from_balance = self.balance_of(*from);
+            if from_balance < value {
+                return Err(PSP22Error::InsufficientBalance);
+            }
+
+            self.balances.insert(from, &(from_balance - value));
+            let to_balance = self.balance_of(*to);
+            self.balances.insert(to, &(to_balance + value));
```

Comment:
should we do arithmetic checks here?

Reply:
not worth the gas cost increase - u128 max is 340282366920938463463374607431768211455, more than the token in circulation.

Reply:
fair point - I would then add a comment about it; I guess that the first vulnerability pattern that the students will learn are over/underflows

Reply:
done